_posts/2017-10-20-totw-55.md

```diff
@@ -0,0 +1,107 @@
+---
+title: "Tip of the Week #55: Name Counting and unique_ptr"
+layout: tips
+sidenav: side-nav-tips.html
+permalink: tips/55
+publish: true
+type: markdown
+---
+
+Originally posted Sep 12, 2013
+
+*by Titus Winters [(titus@google.com)](mailto:titus@google.com)*
+
+Updated Oct 20, 2017
+
+Quicklink: **abseil.io/tips/55**
+
+*"Though we may know Him by a thousand names, He is one and the same to us
+all." - Mahatma Gandhi*
+
+Colloquially, a "name" for a value is any value-typed variable (not a pointer,
+nor a reference), in any scope, that holds a particular data value. (For the
+spec-lawyers, if we say "name" we’re essentially talking about lvalues.) Because
+of `std::unique_ptr`'s specific behavioral requirements, we need to make sure
+that any value held in a `std::unique_ptr` only has one name.
+
+It’s important to note that the C++ language committee picked a very apt name
+for `std::unique_ptr`. Any non-null pointer value stored in a `std::unique_ptr`
+must occur in only one `std::unique_ptr` at any time; the standard library is
+designed to enforce this. Many common problems compiling code that uses
+`std::unique_ptr` can be resolved by learning to recognize how to count the
+names for a `std::unique_ptr`: one is OK, but multiple names for the same
+pointer value are not.
+
+Let’s count some names. At each line number, count the number of names alive at
+that point (whether in scope or not) that refer to a `std::unique_ptr`
+containing the same pointer. If you find any line with more than one name for
+the same pointer value, that’s an error!
+
+```c++
+ std::unique_ptr<Foo> NewFoo() {
```

Comment:
this is indented one space and the rest of the example isn't

Reply:
Fixed.